### ISSUE TITLE
Nodes stamped by dom-if are duplicated when imported-template is reattached to DOM

### ISSUE THREAD

**Issue by warpech** [Juicy/imported-template#26]
This is a continuation of to https://github.com/Starcounter/starcounter-include/issues/17, because after some research, I have managed to further isolate the problem to `imported-template`.

The problem is that nodes stamped by `dom-if` are duplicated when the parent node of `imported-template` is detached and then attached to DOM

I have written a test that fails in `imported-template`. I will upload it as a PR.

I think that the problem is between `imported-template` and Polymer, because using *only* Polymer this works as desired: https://jsbin.com/pesece/1/edit?html,js,output)

**Comment by warpech** [comment]
I did not manage to fix this, though I have isolated the problem in a single test. See the above commit.

This causes a problem in the Website app, meaning that it blocks the Launcher roadmap (https://github.com/StarcounterPrefabs/Launcher/issues/261, point 9)

@miyconst could you please take a look? Maybe a new pair of eyes can fix it.

<img width="1422" alt="screen shot 2017-01-17 at 18 47 52" src="https://cloud.githubusercontent.com/assets/566463/22032891/a2353858-dce6-11e6-80e6-b64893d73597.png">

**Comment by warpech** [comment]
I think I know what the problem is. 

`imported-template` uses `juicy-html`'s **clear** method, which only removes `stampedNodes`, but not `scopedNodes` nor `scopelessNodes`.

Overloading this method in `imported-template` using the following code fixes the problem.

I have made some test last month (commit e539d12) without a solution to the problem. Now I think I have the solution for the problem (see the snippet below)

@tomalec, would you have time to check if this can solve the problem?

```
        var clear = JuicyHTMLElement.prototype.clear;
        ImportedTemplatePrototype.clear = function () {
            var childNo,
                scopedNodes = this.scopedNodes,
                len = scopedNodes && scopedNodes.length || 0;

            for (childNo = 0; childNo < len; childNo++) {
                clear.call({
                  stampedNodes: scopedNodes[childNo]
                });
            }
            this.scopedNodes = [];

            clear.call({
              stampedNodes: this.scopelessNodes
            });
            this.scopelessNodes = [];

            // inline HTML
            if (!this.scopedNodes && !this.scopelessNodes) {
              clear.call({
                stampedNodes: this.stampedNodes
              });
              this.stampedNodes = null;
            }
        }
```

Then, you also need to change:


```
                xhtml.scopedNodes = [];
                xhtml.scopelessNodes = [];
                xhtml.clear();
```

to:

```
                xhtml.clear();
```

**Comment by tomalec** [comment]
Nic catch! I'll check it tomorrow morning

**Comment by tomalec** [comment]
> I think that the problem is between imported-template and Polymer, because using only Polymer this works as desired: https://jsbin.com/pesece/1/edit?html,js,output)

It seems your re-pro with only polymer, was checkign wrong thing. Take a look here
https://jsbin.com/gejevohoge/1/edit?html,js,console,output

**Comment by tomalec** [comment]
I don't think your snippet is needed and could help somehow, as at https://github.com/Juicy/imported-template/blob/master/imported-template.html#L103 we are making `stampedNodes` contain all child nodes of stamped fragment therefore it should contain both scoped and scopeless as well.

Also it does not make tests https://github.com/Juicy/imported-template/tree/reattached-fix/test/use-cases/dom-manipulation pass.

**Comment by warpech** [comment]
Yeah, I just was able to reproduce the original problem by running Website + People + SignIn.

My snippet from https://github.com/Juicy/imported-template/issues/26#issuecomment-279549037 indeed does not help with that problem.

@tomalec any idea how to fix the OP?

**Comment by tomalec** [comment]
Ok, now I can reproduce it, I'll dig deeper

**Comment by tomalec** [comment]
I [have reduced](https://github.com/StarcounterSamples/People/compare/d6469f7627a49a8319662f0da6c5493511e2d7c5...Juicy/imported-template/issues/26) it a little bit: 
Website + [People with two simple HTMLs](https://github.com/StarcounterSamples/People/tree/Juicy/imported-template/issues/26) + SignIn

So, the suspects are : `imported-template`, Polymer, People app

still investigating

**Comment by tomalec** [comment]
I have found something that could cause it https://github.com/StarcounterPrefabs/Website/issues/46, however `imported-template` should handle it anyway

**Comment by tomalec** [comment]
After some investigation I would bet it's related to this issue https://github.com/Polymer/polymer/issues/3682
And I would really like not to try fixing it or working it around in yet another place before Polymer update.

I suspect fixing https://github.com/StarcounterPrefabs/Website/issues/46 should hide this problem for some time.

@warpech WDYT?

**Comment by un-tone** [comment]
Fixing https://github.com/StarcounterPrefabs/Website/issues/46 does not hide the problem.

**Comment by tomalec** [comment]
Could you provide a build with that fix, so I could investigate further?

**Comment by un-tone** [comment]
@tomalec this PR https://github.com/StarcounterSamples/Website/pull/52 hides the problem

**Comment by warpech** [comment]
Get back to this in Starcounter 2.4
